### PR TITLE
Allow manually-awarded experience points to be manually deleted

### PR DESCRIPTION
**Pull request by kxmbrian**

### CODE REVIEW

File: spec/features/course/experience_points_record_management_spec.rb [hunk 1]
```diff
@@ -7,10 +7,17 @@
...
     let(:course_student) { create(:course_student, :approved, course: course) }
-    let!(:record) { create(:course_experience_points_record, course_user: course_student) }
-    let!(:inactive_record) do
+    let(:manual_record) { create(:course_experience_points_record, course_user: course_student) }
+    let(:inactive_record) do
       create(:course_experience_points_record, :inactive, course_user: course_student)
     end
+    let(:record) do
+      points_record = create(:course_experience_points_record, course_user: course_student)
+      points_record.actable = points_record
```
Comment (kxmbrian) [inline spec/features/course/experience_points_record_management_spec.rb]: This is to make it a non-manually-awarded points record. Any concerns if I do this?

File: spec/controllers/course/experience_points_records_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        delete :destroy, course_id: course, course_user_id: course_student, id: points_record_stub
+      end
+
+      context 'when destroy fails' do
+        before do
+          controller.instance_variable_set(:@experience_points_record, points_record_stub)
+          subject
+        end
+
+        it do
```
Comment (lowjoel) [inline spec/controllers/course/experience_points_records_controller_spec.rb]: If you are using a do block, specify the message explicitly.

File: spec/features/course/experience_points_record_management_spec.rb (hunk 1 above)
Comment (lowjoel) [inline spec/features/course/experience_points_record_management_spec.rb]: can't you stub the method out instead?

File: spec/features/course/experience_points_record_management_spec.rb [hunk 3]
```diff
@@ -18,20 +25,40 @@
       let(:user) { create(:course_manager, :approved, course: course).user }
 
       scenario "I can view a course student's active experience points records" do
+        records
```
Comment (lowjoel) [inline spec/features/course/experience_points_record_management_spec.rb]: @allenwq how should we handle variables like this?

File: app/controllers/course/experience_points_records_controller.rb [hunk 4]
```diff
@@ -8,4 +8,16 @@ def index # :nodoc:
     @experience_points_records =
       @experience_points_records.active.includes(creator: :course_users)
   end
+
+  def destroy # :nodoc:
```
Comment (lowjoel) [inline app/controllers/course/experience_points_records_controller.rb]: break up the destroy action into private destroy_success and destroy_failure methods.

File: spec/features/course/experience_points_record_management_spec.rb (hunk 1 above)
Comment (kxmbrian) [inline spec/features/course/experience_points_record_management_spec.rb]: If I stub it like `allow(stub).to receive(:manually_awarded?).and_return(false)`, `@experience_points_record` will not return false in the controller and the view will render incorrectly. Perhaps I could do the same thing (i.e. let a record act as itself), but shift the logic to factory girl?

Comment (lowjoel) [inline spec/features/course/experience_points_record_management_spec.rb]: shifting logic to factory girl is good, thanks.

File: app/models/course/experience_points_record.rb [hunk 5]
```diff
@@ -17,8 +17,6 @@ def active?
     points_awarded.present?
   end
 
-  protected
```
Comment (lowjoel) [inline app/models/course/experience_points_record.rb]: fix the spec for the experience points record also.

File: spec/controllers/course/experience_points_records_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  let!(:instance) { create(:instance) }
+
+  with_tenant(:instance) do
+    let(:course) { create(:course) }
+    let(:course_student) { create(:course_student, :approved, course: course) }
+    let(:user) { create(:course_teaching_assistant, :approved, course: course).user }
+    let(:experience_points_history_path) do
+      course_course_user_experience_points_records_path(course, course_student)
+    end
+    let!(:points_record_stub) do
```
Comment (lowjoel) [inline spec/controllers/course/experience_points_records_controller_spec.rb]: why do you need let!

**Comment by lowjoel** [comment]
@kxmbrian I think this needs a rebase after #899

**Comment by kxmbrian** [comment]
Depends on #909.

**Comment by lowjoel** [comment]
@kxmbrian has this been rebased?

**Comment by kxmbrian** [comment]
@lowjoel Yep, just did a fresh rebase.

File: app/views/course/experience_points_records/_experience_points_record.html.slim [hunk 7]
```diff
@@ -2,3 +2,5 @@
   td = experience_points_record.points_awarded
   td = format_inline_text(experience_points_record.reason)
   td = link_to_user(experience_points_record.creator)
+  - if experience_points_record.manually_awarded? && can?(:destroy, experience_points_record)
```
Comment (lowjoel) [inline app/views/course/experience_points_records/_experience_points_record.html.slim]: Do we want to have this check?

Should we just block all destroys of manually awarded experience points records instead?

**Pull request merged**